rule/defer.go

```diff
@@ -0,0 +1,139 @@
+type DeferRule struct{}
+
+// Apply applies the rule to given file.
+func (r *DeferRule) Apply(file *lint.File, arguments lint.Arguments) []lint.Failure {
+	const disableArgKey = "disable"
+	disable := r.disableFromArgs(arguments)
+
+	var failures []lint.Failure
+	onFailure := func(failure lint.Failure) {
+		failures = append(failures, failure)
+	}
+
+	w := lintDeferRule{onFailure: onFailure, disable: disable}
+
+	ast.Walk(w, file.AST)
+
+	return failures
+}
+
+// Name returns the rule name.
+func (r *DeferRule) Name() string {
+	return "defer"
+}
+
+func (r *DeferRule) disableFromArgs(args lint.Arguments) map[string]bool {
+	result := make(map[string]bool, len(args))
+	if len(args) < 1 {
+		return result
+	}
+
+	aa, ok := args[0].(map[string]interface{})
+	if !ok {
+		panic(fmt.Sprintf("Invalid argument '%v' to the 'defer' rule. Expecting a \"disable\":[]string, got %T", args[0], args[0]))
+	}
+	for k, v := range aa {
+		switch k {
+		case "disable":
+			list, ok := v.([]interface{})
+			if !ok {
+				panic(fmt.Sprintf("Invalid value for \"disable\" argument in 'defer' rule, expected []string, got '%v' (%T)", v, v))
+			}
+			for _, subcase := range list {
+				sc, ok := subcase.(string)
+				if !ok {
+					panic(fmt.Sprintf("Invalid value %v in \"disable\" argument in 'defer' rule, expected string, got (%T)", v, v))
+				}
+				result[sc] = true
+			}
+		default:
+			panic(fmt.Sprintf("Unknown argument %s for 'defer' rule", v))
+		}
+	}
+
+	return result
+}
+
+type lintDeferRule struct {
+	onFailure  func(lint.Failure)
+	inALoop    bool
+	inADefer   bool
+	inAFuncLit bool
+	disable    map[string]bool
+}
+
+func (w lintDeferRule) Visit(node ast.Node) ast.Visitor {
+	switch n := node.(type) {
+	case *ast.ForStmt:
+		w.visitSubtree(n.Body, w.inADefer, true, w.inAFuncLit)
+		return nil
+	case *ast.RangeStmt:
+		w.visitSubtree(n.Body, w.inADefer, true, w.inAFuncLit)
+		return nil
+	case *ast.FuncLit:
+		w.visitSubtree(n.Body, w.inADefer, false, true)
+		return nil
+	case *ast.ReturnStmt:
+		if len(n.Results) != 0 && w.inADefer && w.inAFuncLit {
+			w.newFailure("return in a defer function has no effect", n, 1.0, "logic", "return")
+		}
+	case *ast.CallExpr:
+		if isIdent(n.Fun, "recover") && !w.inADefer {
+			w.newFailure("recover must be called inside a deferred function", n, 1.0, "logic", "recover")
+		}
+	case *ast.DeferStmt:
+		w.visitSubtree(n.Call, true, false, false)
+
+		if w.inALoop {
+			w.newFailure("prefer not to defer inside loops", n, 1.0, "bad practice", "loop")
+		}
+
+		switch fn := n.Call.Fun.(type) {
+		case *ast.CallExpr:
+			w.newFailure("prefer not to defer chains of function calls", fn, 1.0, "bad practice", "call-chain")
+		case *ast.SelectorExpr:
+			if id, ok := fn.X.(*ast.Ident); ok {
+				isMethodCall := id != nil && id.Obj != nil && id.Obj.Kind == ast.Typ
+				if isMethodCall {
+					w.newFailure("be careful when deferring calls to methods without pointer receiver", fn, 0.8, "bad practice", "method-call")
```

Comment:
In what cases can we return false positive?

Reply:
The rule only knows that it is a call to a method but it has no info about the method receiver (pointer or not)
If the receiver is a pointer receiver then deferring is safe

Example:

```go
func (mt myType) foo() {...}
func (mt *myType) bar() {...}
...
{
...
defer foo() // revive alerts 
defer bar() // revive alerts but deferring it is without risk because bar has a pointer receiver
...
}
```